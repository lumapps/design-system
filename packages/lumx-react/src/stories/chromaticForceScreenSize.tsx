/* eslint-disable */
import React from 'react';
import isChromatic from 'chromatic/isChromatic';

/** Story decorator used to force a minimum screen size when running in chromatic */
export const chromaticForceScreenSize = (story: any) =>
    isChromatic() ? <div style={{ minWidth: 1200, minHeight: 800 }}>{story()}</div> : story();
