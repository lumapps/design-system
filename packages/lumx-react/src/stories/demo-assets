../../../site-demo/static/demo-assets