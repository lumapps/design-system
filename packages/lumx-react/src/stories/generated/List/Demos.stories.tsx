/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/list/Demos' };

export { App as Big } from './big';
export { App as Huge } from './huge';
export { App as Regular } from './regular';
export { App as Tiny } from './tiny';
