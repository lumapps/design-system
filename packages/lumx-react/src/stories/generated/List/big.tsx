../../../../../site-demo/content/product/components/list/react/big.tsx