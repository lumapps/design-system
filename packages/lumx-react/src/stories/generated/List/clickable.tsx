../../../../../site-demo/content/product/components/list/react/clickable.tsx