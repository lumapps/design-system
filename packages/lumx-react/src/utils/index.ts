/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
