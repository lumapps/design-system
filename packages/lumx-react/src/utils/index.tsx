export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
