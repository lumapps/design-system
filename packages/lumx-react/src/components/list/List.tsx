import { Size } from '@lumx/react';

import { useKeyboardListNavigation } from '@lumx/react/hooks/useKeyboardListNavigation';
import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';

import classNames from 'classnames';
import React, { forwardRef, Key, ReactNode, SyntheticEvent, useRef } from 'react';
import { useInteractiveList } from './useInteractiveList';

/**
 * Defines the props of the component.
 */
export interface ListProps extends GenericProps {
    /** List content (should be ListItem, ListSubheader or ListDivider). */
    children: ReactNode;
    /**
     * Whether the list items are clickable.
     * @deprecated not needed anymore.
     */
    isClickable?: boolean;
    /** Item padding size. */
    itemPadding?: Extract<Size, 'big' | 'huge'>;
    /**
     * On list item selected callback.
     *
     * @param key   React key of the selected item.
     * @param index Index of the selected item among the sibling items.
     * @param evt   Source event (either mouse or keyboard event).
     */
    onListItemSelected?(key: Key, index: number, evt: SyntheticEvent): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'List';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
/**
 * List component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
const InternalList: Comp<ListProps, HTMLUListElement> = forwardRef((props, ref) => {
    const { children, className, isClickable, itemPadding, onListItemSelected, ...forwardedProps } = props;
    const listElementRef = useRef<HTMLUListElement>(null);

    const { items, hasClickableItem } = useInteractiveList({
        items: children,
        ref: listElementRef,
        onListItemSelected,
    });
    const clickable = hasClickableItem || isClickable;

    return (
        <ul
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                    itemPadding: itemPadding ?? (clickable ? Size.big : undefined),
                }),
            )}
            tabIndex={clickable ? 0 : -1}
            ref={mergeRefs(ref, listElementRef)}
        >
            {items}
        </ul>
    );
});
InternalList.displayName = COMPONENT_NAME;
InternalList.className = CLASSNAME;

export const List = Object.assign(InternalList, { useKeyboardListNavigation });
