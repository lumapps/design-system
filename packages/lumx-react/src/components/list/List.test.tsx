import React from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { commonTestsSuite, itShouldRenderStories } from '@lumx/react/testing/utils';

import { List, ListProps } from './List';
import * as stories from './List.stories';

const CLASSNAME = List.className as string;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (props: Partial<ListProps> = {}, shallowRendering = true) => {
    const renderer: any = shallowRendering ? shallow : mount;
    const wrapper: any = renderer(<List {...(props as any)} />);
    return { props, wrapper };
};

describe(`<${List.displayName}>`, () => {
    // 1. Test render via snapshot.
    describe('Snapshots and structure', () => {
        itShouldRenderStories(stories, List);
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
