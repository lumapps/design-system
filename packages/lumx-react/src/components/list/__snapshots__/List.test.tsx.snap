// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List> Snapshots and structure should render story 'AsLink' 1`] = `
<ul
  className="lumx-list lumx-list--item-padding-big"
  tabIndex={0}
>
  <ListItem
    before={
      <Icon
        icon="M12,4A4,4 0 0,1 16,8A4,4 0 0,1 12,12A4,4 0 0,1 8,8A4,4 0 0,1 12,4M12,14C16.42,14 20,15.79 20,18V20H4V18C4,15.79 7.58,14 12,14Z"
      />
    }
    className="lumx-color-background-dark-L6"
    isHighlighted={false}
    key=".0"
    linkProps={
      Object {
        "href": "#",
      }
    }
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    <span>
      My first link
    </span>
  </ListItem>
  <ListItem
    after={
      <Icon
        icon="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
      />
    }
    className="lumx-color-background-dark-L6"
    isHighlighted={false}
    key=".1"
    linkProps={
      Object {
        "href": "http://www.google.com",
        "target": "_blank",
      }
    }
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    <span>
      Google
    </span>
  </ListItem>
</ul>
`;

exports[`<List> Snapshots and structure should render story 'KeyboardNavigation' 1`] = `
<ul
  className="lumx-list lumx-list--item-padding-big"
  tabIndex={0}
>
  <ListItem
    isHighlighted={false}
    key=".0"
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    Clickable item 1
  </ListItem>
  <ListDivider
    key=".1"
  />
  <ListItem
    isHighlighted={false}
    key=".2"
    linkProps={
      Object {
        "href": "#",
      }
    }
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    Link item 1
  </ListItem>
  <ListItem
    key=".3..0"
    size="regular"
  >
    Non clickable (=non navigable) item
  </ListItem>
  <ListSubheader
    key=".3..1"
  >
    Header
  </ListSubheader>
  <ListItem
    isHighlighted={false}
    key=".3..2"
    linkProps={
      Object {
        "href": "#",
      }
    }
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    Link item 2
  </ListItem>
  <ListSubheader
    key=".3..3"
  >
    Header
  </ListSubheader>
  <CustomListItem
    key=".4"
  />
  <ListItem
    isHighlighted={false}
    key=".5:$1"
    linkProps={
      Object {
        "href": "#",
      }
    }
    linkRef={[Function]}
    onFocus={[Function]}
    onItemSelected={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    size="regular"
  >
    Link item 3
  </ListItem>
</ul>
`;

exports[`<List> Snapshots and structure should render story 'WithCustomChildren' 1`] = `
<ul
  className="lumx-list"
  tabIndex={-1}
>
  <CustomListItem
    key=".0"
  />
  <CustomListItem
    key=".1"
  />
  <CustomListItem
    key=".2"
  />
</ul>
`;

exports[`<List> Snapshots and structure should render story 'WithItemPadding' 1`] = `
Array [
  <ul
    className="lumx-list lumx-list--item-padding-big"
    tabIndex={0}
  >
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".0"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".1"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
  </ul>,
  <ul
    className="lumx-list lumx-list--item-padding-big"
    tabIndex={0}
  >
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".0"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".1"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
  </ul>,
  <ul
    className="lumx-list lumx-list--item-padding-huge"
    tabIndex={0}
  >
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".0"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
    <ListItem
      className="lumx-color-background-dark-L6"
      isHighlighted={false}
      key=".1"
      linkProps={
        Object {
          "href": "#",
        }
      }
      linkRef={[Function]}
      onFocus={[Function]}
      onItemSelected={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      size="big"
    >
      <div>
        Two-line item
      </div>
      <div
        className="lumx-color-font-dark-L2"
      >
        Secondary text
      </div>
    </ListItem>
  </ul>,
]
`;
